sonata-project/SonataPageBundle
# Upload files via Block editor in Page Composer

I am targeting this branch, because this is BC.

Closes #972

## Changelog


```markdown
### Added
- Upload files via block editor in page composer
```

## Subject

With this PR it will be possible to upload files via block editor in page composer. Block editor's form is retargeted into dynamically added iframe instead of sending data with ajax.

To test this feature with existing application:
1. Copy [updated composer.js](https://github.com/covex-nn/SonataPageBundle/blob/page-composer-upload/src/Resources/assets_src/src/js/composer.js) and [current treeview.js](https://github.com/sonata-project/SonataPageBundle/blob/3.x/src/Resources/assets_src/src/js/treeview.js) into `public/issue-972` directory (where `public` - is web root, and `issue-972` - new directory)
2. Update PageBundle configuration:
    ```yaml
    sonata_admin:
        assets:
            extra_javascripts:
                - issue-972/composer.js
                - issue-972/treeview.js
    #            - bundles/sonatapage/sonata-page.back.min.js
            extra_stylesheets:
                - bundles/sonatapage/sonata-page.back.min.css
    ```
3. Create a block with `Sonata\MediaBundle\Form\Type\MediaType` field, or apply some of changes in commit `Demo block` in `upload` branch or my test application https://github.com/covex-nn/sonata-page-app/commits/upload (there are some additional actions/configuration required, and in my application everything works as expected):

    ```php
    use Sonata\AdminBundle\Form\FormMapper;
    use Sonata\BlockBundle\Model\BlockInterface;
    use Sonata\CoreBundle\Form\Type\ImmutableArrayType;
    use Sonata\MediaBundle\Form\Type\MediaType;

    public function buildEditForm(FormMapper $formMapper, BlockInterface $block)
    {
        $this->loadBackground($block);

        $formMapper->add('settings', ImmutableArrayType::class, [
            'keys' => [
                ['background', MediaType::class, [
                    'provider' => 'sonata.media.provider.image',
                    'required' => false,
                    'context' => 'default',
                ]]
            ],
            'translation_domain' => 'AppBlocks',
        ]);
    }
    ```
